docs/_drafts/go-strategy-pattern-a-response.md

```diff
@@ -0,0 +1,172 @@
+Rather than creating an ever-growing switch statement to select between the alternatives, each solution is represented by a class and one such class is passed into the logic that needs to call it.
+In other words, this is an application of **dependency injection**.
+The key thing is that all of these classes, called **strategies**, conform to a common interface that, canonically, exposes a single method to invoke the strategy.
+
+Let's take the example of sorting a list of items into ascending (or descending) order.
+There's quick-sort, merge-sort, heap-sort, insertion-sort, and so on as generally applicable approaches, and specialised algorithms like radix-sort and bucket-sort if dealing with integers.
+Each one can be represented as a single function that accepts a list of some type as input and returns a list of the same type as output.
+If you have some logic that requires sorting a list, you can pass in an appropriate sorting algorithm to that logic, rather than having to hard-code it or embed the decision for which algorithm to apply.
+If you're dealing with `int32`s, you may choose radix-sort; if the list is likely to be almost sorted, you might prefer insertion-sort to quick-sort.
+
+There are three benefits to using the Strategy Pattern.
+Firstly, it decouples the calling logic from the strategy logic, which provides loose coupling and generality, and allows the implementation of incidental logic to be swapped out, such as for testing purposes.
+If the calling context doesn't need to know _how_ something is achieved, but just that it is, then it doesn't need to take responsilbity for that or concern itself with configuration for this; this aligns well with the Single Responsibility Principle.
+Secondly, it provides extensibility in line with the Open-Closed Principle, which is about keeping code more maintainable.
+Thirdly, it keeps the logic grouped into smaller, more focused elements rather than lengthy, expansive blocks.
+That should make it more legible and thus more maintainable, particularly as it's capturing the _purpose_ of these strategies and not just their implementations.[1]
+
+## OOPs!  Go is classless
+
+The description of the Strategy Pattern given above talks about classes, but Go doesn't have classes so how does it apply?!
+Redowan mentions that "the Go community exhibits a knee-jerk reaction to the word 'pattern'" but, rightly in my opinion, continues to use the established term rather than inventing a new one.
+
+What I think is missed here is that Go isn't a million miles away from traditional object-oriented (OO) languages, particularly given the trend of favouring composition over inheritance.
+Crucially, Go supports methods on structs rather than forcing the use of functions without receivers (`func (a *A) foo()` vs `func foo(a *A)`) --- **structs with methods are analogous to classes without subtype polymorphism** (inheritance)!
```

Comment:
:thought_balloon: The examples should probably be parenthetical information at the point of mention rather than at the end of the clause

Reply:
Fixed in https://github.com/agrski/agrski.github.io/pull/45/commits/cd118dd1b3d8b489a741a32e12fa3dd556f576c3